server.js

```diff
@@ -13,8 +13,64 @@ app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
-// TODO add your routes and helper functions here
+app.get("/bookings", function (req, res) {
+  res.send(bookings);
+});
+
+app.get("/bookings/search", function (req, res) {
+  if (req.query.date) {
+    const date = dateToNumber(req.query.date);
+    res.send(bookings.filter((booking) => dateToNumber(booking.checkInDate) <= date && dateToNumber(booking.checkOutDate) >= date));
+  } else {
+    const term = req.query.term;
+    res.send(bookings.filter((booking) => booking.firstName.includes(term) || booking.surname.includes(term) || booking.email.includes(term)));
+  }
+});
+
+app.get("/bookings/:id", function (req, res) {
+  const matched = bookings.find((booking) => booking.id === +req.params.id);
+  if (!matched) return res.status(404).send("incorrect id");
+  res.send(matched);
+});
+
+app.post("/bookings", function (req, res) {
+  const newBooking = req.body;
+  newBooking.id = bookings.length + 1;
+  const { title, firstName, surname, email, roomId, checkInDate, checkOutDate } = newBooking;
+  const valid =
+    !!title &&
+    !!firstName &&
+    !!surname &&
+    !!email &&
+    (roomId === 0 || !!roomId) &&
+    !!checkInDate &&
+    !!checkOutDate &&
+    emailValidation(email) &&
+    dateValidation(dateToNumber(checkInDate), dateToNumber(checkOutDate));
+  if (!valid) return res.status(400).send("missing or incorrect information");
+  bookings.push(newBooking);
+  res.send(bookings);
+});
+
+app.delete("/bookings/:id", (req, res) => {
+  const bookingIndex = bookings.findIndex((booking) => booking.id === +req.params.id);
+  if (bookingIndex === -1) return res.status(404).send("incorrect id");
+  bookings.splice(bookingIndex, 1);
+  res.send(bookings);
```

Comment:
it's not a good idea to send back all bookings, it may contain lot of elements, if we need this response we can use the GET /bookings